### PR TITLE
Add io.github.jakubmelka.pdf4qt

### PR DESCRIPTION
**Pull request by jointri**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. [**Link:**](https://github.com/JakubMelka/PDF4QT/issues/23)
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

This is a huge work in progress, just thought I make the PR now while I fix the flatpak build files. PDF4QT is a very nice PDF reader written from scratch and I am making an effort to get this on flatpak. You can check out upstream here: https://github.com/JakubMelka/PDF4QT

### CODE REVIEW

File: io.github.jakubmelka.pdf4qt.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+app-id: io.github.jakubmelka.pdf4qt
+runtime: org.kde.Platform
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.yaml:2]: you need a runtime-version

File: io.github.jakubmelka.pdf4qt.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+sdk: org.kde.Sdk
+command: pdf4qtbin.sh
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.yaml:4]: you are missing finish-args to at least give the permission to display something

File: io.github.jakubmelka.pdf4qt.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    build-commands: 
+      - install -D pdf4qtbin.sh /app/bin/pdf4qt.sh
+    sources:
+      - type: file
+        path: pdf4qt.sh
+        
+  - name: qbs
+  sources:
+    - type: archive
+      url: https://download.qt.io/official_releases/qbs/1.23.0/qbs-src-1.23.0.tar.gz
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.yaml:17]: need a sha256sum

**Comment by jointri** [comment]
bot, build io.github.jakubmelkda.pdf4qt

File: io.github.jakubmelka.pdf4qt.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  - name: pdf4qtbin
+    buildsystem: simple
+    build-commands: 
+      - install -D pdf4qtbin.sh /app/bin/pdf4qt.sh
+    sources:
+      - type: file
+        path: pdf4qt.sh
+        
+  - name: qbs
+  sources:
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.yaml:15]: this is indented at the wrong level

File: io.github.jakubmelka.pdf4qt.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    - type: archive
+      url: https://download.qt.io/official_releases/qbs/1.23.0/qbs-src-1.23.0.tar.gz
+      commands:
+        - qmake QBS_INSTALL_PREFIX=/app qbs.pro
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.yaml:19]: there is a `buildsystem: qmake`

File: pdf4qtbin.sh [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+
+#All this does is just exec whatever is the argument for the graphical program to be ran.
+exec /app/bin/$1
```
Comment (hfiguiere) [inline pdf4qtbin.sh:4]: this doesn't pass the arguments.
Also is this necessary?

File: io.github.jakubmelka.pdf4qt.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        
+        
+  - name: pdf4qt
+    buildsystem: simple
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.yaml:23]: there is no build-command so nothing is built

**Review by hfiguiere — CHANGES_REQUESTED** [review]
Maybe try to build locally beforehand.

File: pdf4qtbin.sh (hunk 6 above)
Comment (jointri) [reply pdf4qtbin.sh:4]: I wasn't sure how to have multiple binaries in the same flatpak, so I decided to try to make a shell wrapper that would be the entry point. Do you know a better way?

Comment (hfiguiere) [reply pdf4qtbin.sh:4]: the default executable is set in `command`. I don't understand why you need the wrapper script.

File: io.github.jakubmelka.pdf4qt.yaml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+      - type: file
+        path: pdf4qtbin.sh
+        
+  - name: qbs
+    buildsystem: qmake
+    sources:
+    - type: archive
+      url: https://download.qt.io/official_releases/qbs/1.23.0/qbs-src-1.23.0.tar.gz
+      sha256: c86aa775446aec728bcbbed782ec128f4e6e2c26536710017343e684bb616d7a
+      build-commands:
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.yaml:24]: this indentation is incorrect.

File: pdf4qtbin.sh (hunk 6 above)
Comment (jointri) [reply pdf4qtbin.sh:4]: I need a bit of help with this one. I'm not sure what to make the default executable with pdf4qt. The available binaries are: UnitTests, PdfTools, Pdf4QtViewerProfi, Pdf4QtViewerLite, Pdf4QtDocPageOrganizer, and Pdf4QtDocDiff. Which one should I make the default?

**Comment by jointri** [comment]
bot, build io.github.jakubmelkda.pdf4qt

**Comment by jointri** [comment]
bot, build io.github.jakubmelka.pdf4qt

**Comment by jointri** [comment]
bot, build io.github.jakubmelka.pdf4qt

**Comment by jointri** [comment]
bot, stop io.github.jakubmelka.pdf4qt

**Comment by jointri** [comment]
bot, Stop io.github.jakubmelka.pdf4qt

File: io.github.jakubmelka.pdf4qt.json [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        "--socket=x11",
+        "--socket=wayland"
+    ],
+    "modules": [
+        {
+            "name": "qbs",
+            "buildsystem": "qmake",
+            "build-commands": [
+                "qmake QBS_INSTALL_PREFIX=/app qbs.pro"
+            ],
```
Comment (gasinvein) [inline io.github.jakubmelka.pdf4qt.json:16]:
```suggestion
            "buildsystem": "cmake-ninja",
            "builddir": true,
```

File: io.github.jakubmelka.pdf4qt.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    "sdk": "org.kde.Sdk",
+    "finish-args": [
+        "--socket=x11",
+        "--socket=wayland"
+    ],
+    "modules": [
+        {
+            "name": "qbs",
+            "buildsystem": "cmake-ninja",
+            "builddir": true,
```
Comment (gasinvein) [inline io.github.jakubmelka.pdf4qt.json:14]:
```suggestion
            "builddir": true,
            "config-opts": [
                "-DWITH_TESTS:BOOL=OFF"
            ],
            "cleanup": [
                "*"
            ],
```

File: io.github.jakubmelka.pdf4qt.json [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+                {
+                    "type": "archive",
+                    "url": "https://download.qt.io/official_releases/qbs/1.23.0/qbs-src-1.23.0.tar.gz",
+                    "sha256": "c86aa775446aec728bcbbed782ec128f4e6e2c26536710017343e684bb616d7a"
+                }
+            ]
+        },
+        {
+            "name": "PDF4QT",
+            "buildsytem": "simple",
```
Comment (gasinvein) [inline io.github.jakubmelka.pdf4qt.json:31]: And here is the typo:
```suggestion
            "buildsystem": "simple",
```

**Comment by jointri** [comment]
bot, build io.github.jakubmelka.pdf4qt

**Comment by jointri** [comment]
bot, build io.github.jakubmelka.pdf4qt

File: pdf4qtbin.sh (hunk 6 above)
Comment (hfiguiere) [reply pdf4qtbin.sh]: you tell me. I don't know your app.

File: io.github.jakubmelka.pdf4qt.json [hunk 12]
```diff
@@ -0,0 +1,52 @@
+{
+    "id": "io.github.jakubmelka.pdf4qt",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.15-21.08",
+    "sdk": "org.kde.Sdk",
+    "finish-args": [
+        "--socket=x11",
+        "--socket=wayland"
+    ],
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.json:9]:
```suggestion
        "--share=ipc",
        "--socket=fallback-x11",
        "--socket=wayland"
    ],
```

This is needed. Also if wayland is supported, x11 should just be fallback.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by jointri** [comment]
bot, build io.github.jakubmelka.pdf4qt

File: io.github.jakubmelka.pdf4qt.json [hunk 13]
```diff
@@ -0,0 +1,45 @@
+{
+	"id": "io.github.jakubmelka.pdf4qt",
+	"runtime": "org.kde.Platform",
+	"runtime-version": "5.15-21.08",
+	"sdk": "org.kde.Sdk",
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.json:5]: it is still missing `command`

**Comment by hfiguiere** [comment]
bot, build io.github.jakubmelka.pdf4qt

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by jointri** [comment]
bot, build io.github.jakubmelka.pdf4qt

**Comment by jointri** [comment]
bot, build io.github.jakubmelka.pdf4qt

File: io.github.jakubmelka.pdf4qt.json [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+	"command": "pdf4qt",
+	"finish-args": [
+		"--share=ipc",
+		"--socket=fallback-x11",
+		"--socket=wayland"
+	],
+	"modules": [{
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.json:12]:
```suggestion
	"modules": [
	    {
```

File: io.github.jakubmelka.pdf4qt.json [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+			],
+			"sources": []
+		},
+		{
+			"name": "python3-tqdm",
+			"buildsystem": "simple",
+			"build-commands": [
+				"pip3 install --verbose --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} \"tqdm>=4.28.1,<5\" --no-build-isolation"
+			],
+			"sources": [{
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.json:147]:
```suggestion
			"sources": [
			    {
```

File: io.github.jakubmelka.pdf4qt.json [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+				}
+			]
+		},
+		{
+			"name": "python3-Jinja2",
+			"buildsystem": "simple",
+			"build-commands": [
+				"pip3 install --verbose --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} \"Jinja2<3.0.0,>=2.9\" --no-build-isolation"
+			],
+			"sources": [{
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.json:165]:
```suggestion
			"sources": [
			    {
```

File: io.github.jakubmelka.pdf4qt.json (hunk 14 above)
Comment (jointri) [reply io.github.jakubmelka.pdf4qt.json:12]: Sorry about that, it's kinda too late to stop the build, but if it fails again I'll change the braces.

File: io.github.jakubmelka.pdf4qt.json [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+				"sha256": "7f91197cc9e48f989d12e4e6fbc46495c446636dfc81b9ccf50bb0ec74b91d4b"
+			}]
+		},
+		{
+			"name": "python3-Jinja2",
+			"buildsystem": "simple",
+			"build-commands": [
+				"pip3 install --verbose --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} \"Jinja2<4.0.0,>=3.0\" --no-build-isolation"
+			],
+			"sources": [{
```
Comment (hfiguiere) [inline io.github.jakubmelka.pdf4qt.json:195]:
```suggestion
			"sources": [
			    {
```

**Comment by hfiguiere** [comment]
The generated manifest should just be included instead of cut & paste. Otherwise it will be difficult to maintain.

**Comment by jointri** [comment]
I asked upstream if he could move away from conan, if he does so, I can remove the pypi dependencies entirely.

File: io.github.jakubmelka.pdf4qt.json (hunk 14 above)
Comment (hfiguiere) [reply io.github.jakubmelka.pdf4qt.json:12]: it doesn't impact the build, it just impact how readable the JSON is, which is why you had issues before.

File: io.github.jakubmelka.pdf4qt.json [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+			"cleanup": [
+				"*"
+			],
+			"sources": {
+				"type": "archive",
+				"url": "https://download.qt.io/official_releases/qbs/1.23.0/qbs-src-1.23.0.tar.gz",
+				"sha256": "c86aa775446aec728bcbbed782ec128f4e6e2c26536710017343e684bb616d7a"
+			}
+		},
+		{
```
Comment (barthalion) [inline io.github.jakubmelka.pdf4qt.json:28]: Are these generated with one of tools from flatpak-builder-tools? If yes, please move generated dependencies to a separate file; if not, they should be.

**Review by barthalion — APPROVED** [review]

**Review by barthalion — CHANGES_REQUESTED** [review]
Sorry, misclicked. This is of course not approved yet.

**Pull request closed without merging**